core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java

```diff
@@ -1020,6 +1052,87 @@ public void releaseRegisterLease(long workerId) {
     mRegisterLeaseManager.releaseLease(workerId);
   }
 
+  @Override
+  public String getClusterId() {
+    try {
+      // ClusterId is generated by MetaMaster,
+      // MetaMaster is constructed later than BlockMaster,
+      // so wait for MetaMaster to generate and set ClusterId.
+      CommonUtils.waitFor(this + " getClusterId",
+          () -> !mClusterId.get().equals(IdUtils.INVALID_CLUSTER_ID),
+          WaitForOptions.defaults().setTimeoutMs(5000));
+    } catch (Exception e) {
+      LOG.error("getClusterId {}: ", IdUtils.INVALID_CLUSTER_ID, e);
+    }
+    return mClusterId.get();
+  }
+
+  @Override
+  public void setClusterId(String clusterId) {
+    mClusterId.set(clusterId);
+  }
+
+  /**
+   * Check whether the Worker is a member of the current cluster and.
+   * whether it may contain dirty data
+   * @param curClusterId current cluster ID
+   * @param workerClusterId worker report cluster ID
+   * @param blocksNum The number of blocks in the worker
+   * @return instance of RegisterCommandType
+   */
+  private RegisterCommandType checkWorker(String curClusterId, String workerClusterId,
+      int blocksNum) {
+    if (workerClusterId.equals(IdUtils.EMPTY_CLUSTER_ID)) {
+      // A new Worker registers to the Master
+      if (blocksNum == 0) {
+        // The most common case: a new and clean Worker registers to the Master
+        return RegisterCommandType.REGISTER_PERSIST_CLUSTERID;
+      } else {
+        if (ServerConfiguration.getBoolean(PropertyKey.MASTER_CLEAN_DIRTY_WORKER)) {
+          return RegisterCommandType.REGISTER_CLEAN_BLOCKS;
+        } else {
+          // Possible cases:
+          // - A Worker that does not correctly persist the clusterId registers to the Master,
+          //   such as:
+          //   1. In the K8s environment, the file is stored in the memory.
+          //   2. The file that saves clusterId does not have read/write permission
+          //   3. The clusterId file is deleted.
+          //   4. The worker starts for the first time after the upgrade (from
+          //      version “without persist clusterId” upgrade to version “persist clusterId” )
+          //
+          //   In the case 1 ~ 3 the worker will get the default clusterId IdUtils.EMPTY_CLUSTER_ID
+          //   We cannot distinguish whether the worker belongs to the current cluster.
+          //   Although the worker may contain dirty data. But it is an undefined behavior, the
+          //   clusterId file should be be persisted correctly.
+          //   All of these cases are due to clusterId not being persisted correctly the branch
+          //   should not be entered except the case 4.
```

Comment:
This is a great question. I personally prefer deleting these blocks because losing cache is better than having incorrect blocks. WDYT @beinan ?

Reply:
My concern is case 4, some of the users might be not happy to lose all the blocks they cached when they perform the upgrade.

Reply:
yes, In the current design, if it is case4, the user will not lose any data, and the upgrade can be completed normally